File: QuipMac/Services/ConnectionLog.swift
// ConnectionLog.swift
// QuipMac — In-memory ring buffer of recent WebSocket connection events
// Feeds the "Diagnostics" section in Settings so the user can see who
// tried to connect, when, and whether it worked.

import Foundation
import Observation

/// One entry in the connection log.
struct ConnectionEvent: Identifiable, Sendable {
    let id = UUID()
    let timestamp: Date
    let kind: Kind
    /// Best-effort remote address. NWConnection doesn't always give us a tidy
    /// IP — sometimes it's "hostname:port", sometimes a raw endpoint descriptor.
    /// We store whatever we can pull and let the UI render it as-is.
    let remote: String
    let detail: String?

    enum Kind: String, Sendable {
        case connected
        case disconnected
        case authSucceeded
        case authFailed
        case failed
    }
}

@MainActor
@Observable
final class ConnectionLog {
    /// Most recent events first. Capped to keep memory bounded — a phone that
    /// reconnects every few seconds because of a flaky tunnel shouldn't balloon
    /// this into MBs.
    private(set) var events: [ConnectionEvent] = []

    /// 20 is enough history to see the last handful of connect/disconnect/auth
    /// cycles without turning the settings panel into a scroll marathon.
    static let maxEvents = 20

    func record(_ kind: ConnectionEvent.Kind, remote: String, detail: String? = nil) {
        let event = ConnectionEvent(timestamp: Date(), kind: kind, remote: remote, detail: detail)
        events.insert(event, at: 0)
        if events.count > Self.maxEvents {
            events.removeLast(events.count - Self.maxEvents)
        }
    }

    func clear() {
        events.removeAll()
    }
}

File: QuipLinux/src/services/connection_log.rs
use std::collections::VecDeque;
use std::sync::{Arc, Mutex};
use std::time::SystemTime;

/// What kind of connection event was recorded. Mirrors Mac's
/// `ConnectionEvent.Kind`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConnectionEventKind {
    Connected,
    Disconnected,
    AuthSucceeded,
    AuthFailed,
    Failed,
}

impl ConnectionEventKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Connected => "connected",
            Self::Disconnected => "disconnected",
            Self::AuthSucceeded => "auth_succeeded",
            Self::AuthFailed => "auth_failed",
            Self::Failed => "failed",
        }
    }
}

#[derive(Debug, Clone)]
pub struct ConnectionEvent {
    pub timestamp: SystemTime,
    pub kind: ConnectionEventKind,
    pub remote: String,
    pub detail: Option<String>,
}

/// In-memory ring buffer of recent WebSocket connection events. Mirrors
/// `QuipMac/Services/ConnectionLog.swift`. Cap is 20 entries — enough to see
/// the last few connect/disconnect/auth cycles without ballooning into MBs
/// when a flaky tunnel reconnects every few seconds.
#[derive(Default)]
pub struct ConnectionLog {
    inner: Mutex<VecDeque<ConnectionEvent>>,
}

impl ConnectionLog {
    pub const MAX_EVENTS: usize = 20;

    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            inner: Mutex::new(VecDeque::new()),
        })
    }

    pub fn record(&self, kind: ConnectionEventKind, remote: impl Into<String>, detail: Option<String>) {
        let event = ConnectionEvent {
            timestamp: SystemTime::now(),
            kind,
            remote: remote.into(),
            detail,
        };
        let mut buf = self.inner.lock().expect("connection_log poisoned");
        buf.push_front(event);
        while buf.len() > Self::MAX_EVENTS {
            buf.pop_back();
        }
    }

    pub fn snapshot(&self) -> Vec<ConnectionEvent> {
        self.inner
            .lock()
            .expect("connection_log poisoned")
            .iter()
            .cloned()
            .collect()
    }

    pub fn clear(&self) {
        self.inner.lock().expect("connection_log poisoned").clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn record_inserts_at_front() {
        let log = ConnectionLog::new();
        log.record(ConnectionEventKind::Connected, "1.2.3.4:5555", None);
        log.record(ConnectionEventKind::AuthSucceeded, "1.2.3.4:5555", None);
        let events = log.snapshot();
        assert_eq!(events.len(), 2);
        assert_eq!(events[0].kind, ConnectionEventKind::AuthSucceeded);
        assert_eq!(events[1].kind, ConnectionEventKind::Connected);
    }

    #[test]
    fn caps_at_max_events() {
        let log = ConnectionLog::new();
        for i in 0..(ConnectionLog::MAX_EVENTS + 5) {
            log.record(ConnectionEventKind::Connected, format!("client-{i}"), None);
        }
        let events = log.snapshot();
        assert_eq!(events.len(), ConnectionLog::MAX_EVENTS);
        // Newest first — last recorded is at the head.
        assert_eq!(
            events[0].remote,
            format!("client-{}", ConnectionLog::MAX_EVENTS + 4)
        );
    }

    #[test]
    fn clear_empties_buffer() {
        let log = ConnectionLog::new();
        log.record(ConnectionEventKind::Connected, "x", None);
        log.clear();
        assert!(log.snapshot().is_empty());
    }
}
